### PR TITLE
[pollytts] Polly Text-to-Speech initial contribution

### LINKED ISSUE #1439: Add Amazon Polly TTS Voice (previously Ivona Speech Cloud)
As a big enthusiast of openHAB I'd like to suggest implementing yet another Text-to-Speech service, based on [Ivona Speech Cloud](http://developer.ivona.com/en/speechcloud/introduction.html).
It's a polish service (the company was acquired by Amazon some time ago) that produces really accurate and natural sounding voices. And - what's important for me, particularly - has great Polish language support, unlike e.g. VoiceRSS. They have a lot of choice for [other languages as well](https://www.ivona.com/us/).

I've already did some research in that field and even started coding my own add-on (which is cool, because I'm a front-end dev and didn't touch Java for years!)

So in order to use Ivona API, you'll need API credentials (API Key and API Secret)

They developed a SDK to get you started:
https://github.com/IvonaSoftware/ivona-speechcloud-sdk-java

So the Ivona openHAB voice addon will require additional lib.

Let me know what you think on the idea. 
I hope I'll submit some pull request soon.

### PR DESCRIPTION
**Pull request by hillmanr**
Goal is a TTS voice service for Openhab 2.x utilizing a Internet based TTS service provided by amazon called Polly. Will generate speech from both plain text input and Speech Synthesis Markup Language.  Provides multiple languages and multiple voices for each language.

PollyTTS Text-to-Speech

"Initial contribution"

Examples:
say("Hello there")  
say("Hello there", "pollytts:Joanne", "enhancedjavasound")  
say("" + item.state,"pollytts:Joey", "enhancedjavasound")  
say("<speak>Children, come to dinner <prosody volume='x-loud'>Right now!</prosody></speak>")  

Community forum thread
https://community.openhab.org/t/amazon-polly-tts-voice-service/29699

Fixes #1439

### CODE REVIEW

**Comment by wborn** [comment]
Many review comments also seem to apply to the VoiceRSS addon you've used as sample for your code. So I've created PR #3809 to improve the VoiceRSS addon as well!

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImplementation.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+    private void purgeAgedFiles() throws IOException {
+        // just exit if expiration set to 0/disabled
+        if (PollyClientConfig.getExpireDate() == 0) {
+            return;
+        }
+        long now = new Date().getTime();
+        long diff = now - PollyClientConfig.getlastDelete();
+        // only execute ~ once every 2 days if cache called
+        long oneDay = 24 * 60 * 60 * 1000;
```
Comment (wborn) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImplementation.java]: You can use TimeUnit for calculating these values, e.g.:

`TimeUnit.DAYS.toMillis(1)`

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+            // create voice to ID translation for service invocation
+            for (Voice voice : PollyClientConfig.pollyVoices) {
+                labelToID.put(voice.getName(), voice.getId());
+            }
+
+            // Initialize expired check date to 172800000 +1
+            // run today and ~ every 2 days if cache cleaner enabled
+            // 1 day = 24 * 60 * 60 * 1000 = 86,400,000
+            today = 172800001;
```
Comment (wborn) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java]: `TimeUnit.DAYS.toMillis(2) + 1`

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+            initialized = true;
```
Comment (wborn) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java]: There's no need for the initialized variable if you just `return true;` here and `return false;` in the catch.

File: addons/voice/org.openhab.voice.pollytts/META-INF/MANIFEST.MF [hunk 4]
```diff
@@ -0,0 +1,24 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: PollyTSS Voice Service
+Bundle-SymbolicName: org.openhab.voice.pollytts;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.3.0.qualifier
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/META-INF/MANIFEST.MF]: done

File: addons/voice/org.openhab.voice.pollytts/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+Bundle-ClassPath: .,
+ lib/aws-java-sdk-polly-1.11.136.jar,
+ lib/aws-java-sdk-core-1.11.136.jar,
+ lib/commons-logging-1.2.jar,
+ lib/jackson-databind-2.6.6.jar,
+ lib/jackson-core-2.6.6.jar,
+ lib/jackson-annotations-2.6.0.jar,
+ lib/httpclient-4.5.2.jar,
+ lib/httpcore-4.4.4.jar,
+ lib/joda-time-2.8.1.jar
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/META-INF/MANIFEST.MF]: done

File: addons/voice/org.openhab.voice.pollytts/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -0,0 +1,24 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/META-INF/MANIFEST.MF]: done

File: addons/voice/org.openhab.voice.pollytts/META-INF/MANIFEST.MF [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ lib/aws-java-sdk-core-1.11.136.jar,
+ lib/commons-logging-1.2.jar,
+ lib/jackson-databind-2.6.6.jar,
+ lib/jackson-core-2.6.6.jar,
+ lib/jackson-annotations-2.6.0.jar,
+ lib/httpclient-4.5.2.jar,
+ lib/httpcore-4.4.4.jar,
+ lib/joda-time-2.8.1.jar
+Import-Package: 
+ org.eclipse.smarthome.config.core,
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/META-INF/MANIFEST.MF]: Done

File: addons/voice/org.openhab.voice.pollytts/README.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
+# Pollytts Text-to-Speech
+
+## Overview
+
+PollyTTS is a TTS voice service for Openhab 2.x utilizing a Internet based TTS service provided by amazon called Polly.  Will generate speech from both plain text input and Speech Synthesis Markup Language. There are servers set in various geographic locations. It requires a set of API keys provided by Amazon to get access to the service. Provides multiple languages and multiple voices for each language.
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/README.md]: done

File: addons/voice/org.openhab.voice.pollytts/README.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+accessKey - required credential provided by Amazon 
+
+secretKey - required credential provided by Amazon
+
+serviceRegion - Required value select region closest for best response. ServiceRegion is one of the following:  
+["us-east-2"] in US East (Ohio)  
+["us-east-1"] in US East (N. Virginia)  
+["us-west-2"] in US West (Oregon)  
+["eu-west-1"] in EU (Ireland)
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/README.md]: Originally there were only 4 regions.  Updated description to describe user specified field and expanded regions.

File: addons/voice/org.openhab.voice.pollytts/ESH-INF/config/config.xml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+		</parameter>
+		
+		<parameter name="serviceRegion" type="text">
+                    <label>Service Region</label>
+                    <description>The Service Region is need to select which region is closest to you.</description>
+                    <options>
+                       <option value="us-east-2">US East (Ohio)</option>
+                       <option value="us-east-1">US East (N. Virginia)</option>
+                       <option value="us-west-2">US West (Oregon)</option>
+                       <option value="eu-west-1">EU (Ireland)</option>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/ESH-INF/config/config.xml]: modified. Removed options for regions and now a used specified field.

File: addons/voice/org.openhab.voice.pollytts/pom.xml [hunk 11]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns="http://maven.apache.org/POM/4.0.0"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<parent>
+		<groupId>org.openhab.addons.voice</groupId>
+		<artifactId>pom</artifactId>
+		<version>2.3.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/pom.xml]: done

File: addons/voice/org.openhab.voice.pollytts/pom.xml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+			<version>1.11.136</version>
+		</dependency>
+		
+	</dependencies>
+		
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>org.openhab.voice</groupId>
+	<artifactId>org.openhab.voice.pollytts</artifactId>
+
+	<name>PollyTTS TTS Integration</name>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/pom.xml]: done

File: addons/voice/org.openhab.voice.pollytts/pom.xml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+		<dependency>
+			<groupId>com.amazonaws</groupId>
+			<artifactId>aws-java-sdk-polly</artifactId>
+			<version>1.11.136</version>
+		</dependency>
+		
+	</dependencies>
+		
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>org.openhab.voice</groupId>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/pom.xml]: done

File: addons/voice/org.openhab.voice.pollytts/pom.xml [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+	<artifactId>org.openhab.voice.pollytts</artifactId>
+
+	<name>PollyTTS TTS Integration</name>
+
+	<packaging>eclipse-plugin</packaging>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/pom.xml]: done

File: addons/voice/org.openhab.voice.pollytts/pom.xml [hunk 15]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns="http://maven.apache.org/POM/4.0.0"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<parent>
+		<groupId>org.openhab.addons.voice</groupId>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/pom.xml]: done

File: addons/voice/org.openhab.voice.pollytts/pom.xml [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+		<version>2.3.0-SNAPSHOT</version>
+	</parent>
+	
+<dependencies>
+		<!-- https://mvnrepository.com/artifact/com.amazonaws/aws-java-sdk-polly -->
+		<dependency>
+			<groupId>com.amazonaws</groupId>
+			<artifactId>aws-java-sdk-polly</artifactId>
+			<version>1.11.136</version>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/pom.xml]: done

File: addons/voice/org.openhab.voice.pollytts/.classpath [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+	<classpathentry kind="output" path="target/classes"/>
+	<classpathentry kind="lib" path="lib/aws-java-sdk-polly-1.11.136.jar"/>
+	<classpathentry kind="lib" path="lib/aws-java-sdk-core-1.11.136.jar"/>
+	<classpathentry kind="lib" path="lib/commons-logging-1.2.jar"/>
+	<classpathentry kind="lib" path="lib/jackson-databind-2.6.6.jar"/>
+	<classpathentry kind="lib" path="lib/jackson-core-2.6.6.jar"/>
+	<classpathentry kind="lib" path="lib/jackson-annotations-2.6.0.jar"/>
+	<classpathentry kind="lib" path="lib/httpclient-4.5.2.jar"/>
+	<classpathentry kind="lib" path="lib/httpcore-4.4.4.jar"/>
+	<classpathentry kind="lib" path="lib/joda-time-2.8.1.jar"/>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/.classpath]: done

**Comment by wborn** [comment]
Thanks for starting  to process my comments @hillmanr! 

Can you also rebase the branch of this PR?

I think you merged commits using the GitHub UI which has resulted in 100 thousands of changes. 

selection_001

It has become very hard to review the changes in this PR now. :wink:

**Comment by hillmanr** [comment]
Ok. I'm not very good with that could you send me a pointer on doing that.
I had a difficult time last time I tried.

**Comment by wborn** [comment]
I usually use the git command like explained on [Stack Overflow](https://stackoverflow.com/questions/7929369/how-to-rebase-local-branch-with-remote-master).

**Comment by wborn** [comment]
I've rebased this branch on the current master and also addressed all the review comments @hillmanr!
It still works very well so it would be nice if you and some other @openhab/2-x-add-ons-maintainers can check the end result so we can merge this PR. :smile:

File: addons/voice/org.openhab.voice.pollytts/README.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+## Service Configuration
+
+Using your favorite configuration UI (e.g. Paper UI) edit **Services/Voice/Polly Text-to-Speech** settings and set:
+
+* **Access Key** - The AWS credentials access key (required).
+* **Secret Key** - The AWS credentials secret key (required).
+* **Service Region** - The service region used for accessing Polly (required). To reduce latency select the region closest to you. E.g. "eu-west-1" (see [regions](https://docs.aws.amazon.com/general/latest/gr/rande.html#pol_region))
+
+The PollyTTS service caches audio files from previous requests. This reduces traffic, improves performance, reduces the number of requests and provides offline functionality.
```
Comment (Confectrician) [inline addons/voice/org.openhab.voice.pollytts/README.md]: This could be improved to one line per sentence.

Comment (wborn) [reply addons/voice/org.openhab.voice.pollytts/README.md]: It's fixed!

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+        } catch (IllegalArgumentException e) {
+            logger.warn("Failed to initialize PollyTTS: {}", e.getMessage());
+        } catch (Exception e) {
+            logger.warn("Failed to initialize PollyTTS", e);
+        }
+    }
+
+    @Override
+    public Set<Voice> getAvailableVoices() {
+        return voices;
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: We should return an immutable set. Shouldn't we?

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    @Override
+    public Set<AudioFormat> getSupportedFormats() {
+        return audioFormats;
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: We should return an immutable set. Shouldn't we?

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+    private Set<AudioFormat> initAudioFormats() {
+        // @formatter:off
+        return pollyTTSImpl.getAvailableAudioFormats().stream()
+                .map(this::getAudioFormat)
+                .collect(toSet());
+        // @formatter:on
+    }
+
+    private AudioFormat getAudioFormat(String apiFormat) {
+        if ("MP3".equals(apiFormat)) {
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]:
```suggestion
        if (CODEC_MP3.equals(apiFormat)) {
```

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+            // use by default: MP3, 22khz_16bit_mono with bitrate 64 kbps
+            return new AudioFormat(CONTAINER_NONE, CODEC_MP3, null, 16, 64000, 22050L);
+        } else if ("OGG".equals(apiFormat)) {
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]:
```suggestion
        } else if (CONTAINER_OGG.equals(apiFormat)) {
```

Or define an own `static final String` for it.

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+        }
+    }
+
+    private String getApiAudioFormat(AudioFormat format) {
+        if (!"default".equals(pollyTTSConfig.getAudioFormat())) {
+            // Override system specified with user preferred value
+            return pollyTTSConfig.getAudioFormat();
+        }
+        if (CODEC_MP3.equals(format.getCodec())) {
+            return "MP3";
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: see above "MP3"

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+        } else if (CODEC_VORBIS.equals(format.getCodec())) {
+            return "OGG";
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: see above "OGG"

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImpl.java [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+        int read = inputStream.read(bytes, 0, READ_BUFFER_SIZE);
+        while (read > 0) {
+            outputStream.write(bytes, 0, read);
+            read = inputStream.read(bytes, 0, READ_BUFFER_SIZE);
+        }
+    }
+
+    private void writeText(File file, String text) throws IOException {
+        try (OutputStream outputStream = new FileOutputStream(file)) {
+            outputStream.write(text.getBytes("UTF-8"));
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImpl.java]:
```suggestion
            outputStream.write(text.getBytes(StandardCharsets.UTF_8));
```

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImpl.java [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+     * Gets a unique filename for a give text, by creating a MD5 hash of it. It
+     * will be preceded by the voice label.
+     *
+     * Sample: "Robert_00a2653ac5f77063bc4ea2fee87318d3"
+     */
+    private String getUniqueFilenameForText(String text, String label) {
+        byte[] bytesOfMessage;
+        MessageDigest md;
+        try {
+            bytesOfMessage = text.getBytes("UTF-8");
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImpl.java]:
```suggestion
            bytesOfMessage = text.getBytes(StandardCharsets.UTF_8);
```

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyTTSCloudImpl.java [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+ * <li>All voices</li>
+ * <li>MP3 and OGG formats</li>
+ * </ul>
+ *
+ * @author Robert Hillman - Initial contribution
+ */
+public class PollyTTSCloudImpl {
+
+    private static final Set<String> SUPPORTED_AUDIO_FORMATS = Collections
+            .unmodifiableSet(Stream.of("MP3", "OGG").collect(toSet()));
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyTTSCloudImpl.java]: See above "MP3", "OGG"

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImpl.java [hunk 28]
```diff
@@ -0,0 +1,169 @@
...
+            outputStream.write(text.getBytes("UTF-8"));
+        }
+    }
+
+    private void updateTimeStamp(File file) throws IOException {
+        // update use date for cache management
+        file.setLastModified(System.currentTimeMillis());
+    }
+
+    private void purgeAgedFiles() throws IOException {
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImpl.java]: Nice. I will consider this in my approach for https://github.com/eclipse/smarthome/issues/6041.

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 29]
```diff
@@ -0,0 +1,240 @@
...
+
+    @Modified
+    protected void modified(Map<String, Object> config) {
+        try {
+            pollyTTSConfig = new PollyTTSConfig(config);
+            logger.debug("Using configuration {}", config);
+
+            // create cache folder
+            File userData = new File(ConfigConstants.getUserDataFolder());
+            File cacheFolder = new File(new File(userData, CACHE_FOLDER_NAME), SERVICE_PID);
```
Comment (cweitkamp) [inline addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: You do not nee a local variable for `userData`.

```suggestion
            File cacheFolder = new File(new File(new File(ConfigConstants.getUserDataFolder()), CACHE_FOLDER_NAME), SERVICE_PID);
```

Comment (Hilbrand) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: You don't even need the `new File` :smile:

```suggestion
            File cacheFolder = new File(new File(ConfigConstants.getUserDataFolder(), CACHE_FOLDER_NAME), SERVICE_PID);
```

**Comment by hillmanr** [comment]
Thanks for all the work.

**Comment by wborn** [comment]
Thanks for the additional review comments @cweitkamp , @Hilbrand !
These should now also be addressed with the latest commit.

**Pull request merged**